File: electron/server/schemas/preventive.ts
import { z } from "zod";
import { weekYearRegex } from '../../utils/weekTools'

export const workerSchema = z.object({
    id: z.number(),
    registration: z.number(),
    name: z.string(),
    class: z.string(),
})

export const machineSchema = z.object({
    id: z.number(),
    tag: z.string(),
    technology: z.string(),
    ute: z.string()
})

export const natureSchema = z.object({
    id: z.number(),
    name: z.string()
})

export const actionCreateSchema = z.object({
    description: z.string(),
    machineId: z.number(),
    excution: z.string(),
    frequency: z.number(),
    natureId: z.number(),
    nextExecution: z.string().regex(weekYearRegex),
    preventiveOSId: z.number().nullable().optional(),
    ignore: z.boolean()
})

export const actionsSchema = z.object({
    id: z.number(),
    machine: machineSchema.optional(),
    nature: natureSchema.optional(),
    _count: z.object({ actionsTaken: z.number() }).optional(),
    ...actionCreateSchema.shape
})

export const actionsTakenSchema = z.object({
    id: z.number(),
    date: z.date().or(z.string()),
    osId: z.number(),
    actionId: z.number(),
    weekCode: z.string().regex(weekYearRegex),
    action: actionsSchema
})

export const serviceOrdersSchema = z.object({
    id: z.number().optional(),
    concluded: z.boolean().nullable().optional(),
    date: z.date().nullable().optional().or(z.string()),
    machineId: z.number(),
    weekCode: z.string().regex(weekYearRegex),
    natureId: z.number(),
    responsible: z.array(workerSchema).optional(),
    actions: z.array(actionsSchema).optional(),
    actionsUniqueKey: z.string(),
    machine: machineSchema.optional(),
    nature: natureSchema.optional(),
    duration: z.number().optional().nullable(),
    actionsTaken: z.array(actionsTakenSchema).optional(),
    startTime: z.date().optional().nullable(),
    finishTime: z.date().optional().nullable(),
})

export type ServiceOrdersType = z.infer<typeof serviceOrdersSchema>

export const assembleServiceOrdersParamsSchema = z.object({
    week: z.number(),
    year: z.number(),
    status: z.string(),
    nature: z.number(),
    machine: z.number(),
})

export type AssembleServiceOrdersParamsType = z.infer<typeof assembleServiceOrdersParamsSchema>

export const executeServiceOrdersParamsSchema = z.object({
    id: z.number(),
    date: z.string(),
    workers: z.array(z.object({ id: z.number() })),
    startTime: z.string(),
    finishTime: z.string(),
    duration: z.number(),
    IdsOfActionsTaken: z.array(z.object({ id: z.number() })).optional(),
})

export type ExecuteServiceOrdersType = z.input<typeof executeServiceOrdersParamsSchema>

export const generateActionsUniqueKeyParmsSchema = z.object({
    id: z.number().optional(),
    machineId: z.number(),
    natureId: z.number(),
})

export type GenerateActionsUniqueKeyParms = z.infer<typeof generateActionsUniqueKeyParmsSchema>[]

File: src-py/schemas/preventive.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class GetServiceOrderByIdParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)


class Machine(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    tag: str
    ute: str
    technology: str


class Nature(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    name: str


class Worker(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    registration: int
    name: str
    workerClass: str


class Action(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    description: str
    machineId: int
    excution: str
    frequency: int
    nextExecution: str
    preventiveOSId: Optional[int] = None
    natureId: int
    ignore: bool
    machine: Optional[Machine] = None
    nature: Optional[Nature] = None


class ActionTaken(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    date: datetime
    osId: int
    actionId: int
    weekCode: str
    action: Action


class ActionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    description: str
    machineId: int
    excution: str
    frequency: int
    natureId: int
    nextExecution: str
    preventiveOSId: Optional[int] = None
    ignore: bool


class UpdateActionParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)
    data: ActionCreate


class DeleteActionParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)


class WorkerRef(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)


class ExecuteServiceOrderData(BaseModel):
    model_config = ConfigDict(extra="forbid")

    date: str
    workers: list[WorkerRef] = Field(default_factory=list)
    startTime: str
    finishTime: str


class ExecuteServiceOrdersParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)
    date: str
    workers: list[WorkerRef] = Field(default_factory=list)
    startTime: str
    finishTime: str


class UpdateServiceOrderParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)
    data: ExecuteServiceOrderData


class DeleteServiceOrderParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int = Field(gt=0)


class GetServiceOrderCountParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    week: int
    year: int


class GetActionsParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    searchText: str
    weekCode: str
    machineId: int
    natureId: int
    showIgnore: bool
    limit: Optional[int] = None
    cursor: Optional[int] = None


class AssembleServiceOrdersParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    week: int
    year: int
    status: str
    nature: int
    machine: int


class ServiceOrderCount(BaseModel):
    model_config = ConfigDict(extra="ignore")

    finished: int
    unfinished: int


class ServiceOrder(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: int
    weekCode: str
    date: Optional[datetime] = None
    natureId: int
    duration: Optional[int] = None
    concluded: Optional[bool] = None
    startTime: Optional[datetime] = None
    finishTime: Optional[datetime] = None
    machineId: int
    actionsUniqueKey: str
    nature: Optional[Nature] = None
    machine: Optional[Machine] = None
    responsible: list[Worker] = Field(default_factory=list)
    actions: list[Action] = Field(default_factory=list)
    actionsTaken: list[ActionTaken] = Field(default_factory=list)
